### DISCUSSION TITLE
Is it possible to parse big csv file line by line?

### DISCUSSION THREAD

**Discussion by alfonz19** [FasterXML/jackson-dataformats-text#505] [Usage]
Maybe I'm overlooking this option somehow, if so please be so kind and point to related documentation/code.

Scenario: In deserializing to some java dto, csv file with first line being header(potentially invalid not relating to dto), following lines are records, potentially invalid, either for deserialization to dto or bean validation. I'd like to process the whole file and report all lines which contains error.

Currently I'm solving it by reading first line (header) which I store in variable and while reading remaining lines I'm creating 2LOC csv documents, which I parse separately. It's nasty, but I don't see another option. Please advise

**Comment by cowtowncoder** [comment]
Sounds like a discussion, not issue, converting.

**Reply by cowtowncoder** [reply]
Line-by-line reading is done using `MappingIterator`. Hopefully you have read:

https://medium.com/@cowtowncoder/reading-csv-with-jackson-c4e74a15ddc1

If that doesn't help, please show code you have done so far -- how are you reading CSV, if not line by line?

**Reply by alfonz19** [reply]
Thanks for reply, let me show what I mean. Apologies for crude-er code and some nasty constructs in this mwe...

Yes, I read that article and tried to walk through source code/documentation. Base on this I can write say this test:

```public class CsvTest {

    @Test
    void testValid() throws IOException {
        String input = """
                "A","B"
                "valid1", true
                "valid2", false
                """;

        MappingIterator<ImportCsvDataWithBooleanPrimitive> mi = new CsvMapper()
                .readerFor(ImportCsvDataWithBooleanPrimitive.class)
                .with(CsvSchema.emptySchema().withHeader().withArrayElementSeparator("|"))
                .with(CsvParser.Feature.EMPTY_STRING_AS_NULL)
                .with(CsvParser.Feature.FAIL_ON_MISSING_COLUMNS)
                .readValues(new InputStreamReader(new ByteArrayInputStream(input.getBytes(StandardCharsets.UTF_8))));
        List<ImportCsvDataWithBooleanPrimitive> result = mi.readAll();

        assertThat(result, notNullValue());
        assertThat(result.size(), is(2));
    }

    public static final class ImportCsvDataWithBooleanPrimitive {
        @JsonProperty("A")
        private String a;
        @JsonProperty("B")
        private boolean b;
    }
}
```

question 1 here already: there is option `CsvGenerator.Feature#ALWAYS_QUOTE_NUMBERS` and `CsvGenerator.Feature#ALWAYS_QUOTE_EMPTY_STRINGS`, but IIUC some older libraries (which I'm trying to move away from) quotes everything. Is there a possibility to also quote booleans and then parsed them back if they were serialized quoted?

But back to the topic. Invalid sample parsing would look like this:

```
@Test
    void testInvalid() {
        String input = """
                "A","B"
                "valid1", "true"
                "invalid1", "not-so-much1"
                "valid2", "false"
                "invalid2", "not-so-much2"
                """;


        InvalidFormatException exception = assertThrows(InvalidFormatException.class, () -> {
            MappingIterator<ImportCsvDataWithBooleanPrimitive> mi = new CsvMapper()
                    .readerFor(ImportCsvDataWithBooleanPrimitive.class)
                    .with(CsvSchema.emptySchema().withHeader().withArrayElementSeparator("|"))
                    .with(CsvParser.Feature.EMPTY_STRING_AS_NULL)
                    .with(CsvParser.Feature.FAIL_ON_MISSING_COLUMNS)
                    .readValues(new InputStreamReader(new ByteArrayInputStream(input.getBytes(StandardCharsets.UTF_8))));
            mi.readAll();
        });

        assertThat(exception.getMessage(), Matchers.startsWith("Cannot deserialize value of type `boolean` from String \"\"true\"\": only \"true\"/\"True\"/\"TRUE\" or \"false\"/\"False\"/\"FALSE\" recognized"));
    }
```

but you see, we failed on first error, diplaying only this error, since parsing stopped. Ideally I'd like to continue parsing, and provide all errors, otherwise the client might just send it again to get another round or parsing attempt. Ideal response from my code would be: lines 1,3,5 fails to be deserialized(or has too many/too much column), lines 2,4,6 fails bean validation (out of scope here). Or the input/header is empty. I didn't find any way how to initialize header based on first column and then read each line as a new entry, collecting all errors with respective line numbers

So the best shot I was able to do, reading 1-by-1 with first line being header and producing human-readable error messages for each failed lines follows. But it's really ugly and does not seem correct.

```
@Test
    void testInvalid_1By1()  {
        String input = """
                "A","B"
                "valid1", true
                "invalid1", true, "this", "is", "too", "much!"
                "invalid2", "false"
                "invalid3"
                """;

        AtomicReference<List<String>> errors = new AtomicReference<>();
        read1By1(input, errors::set);


        String expected = """
                #3: Too many entries: expected at most 2 (value #2 (6 chars) ""this"")
                #4: Cannot deserialize value of type `boolean` from String ""false"": only "true"/"True"/"TRUE" or "false"/"False"/"FALSE" recognized
                #5: Not enough column values: expected 2, found 1""";

        assertThat(String.join("\n", errors.get()), is(expected));
    }

    private List<ImportCsvDataWithBooleanPrimitive> read1By1(String input, Consumer<List<String>> errorConsumer) {
        List<ImportCsvDataWithBooleanPrimitive> result = new ArrayList<>();
        List<String> parsingErrors = new ArrayList<>();
        ObjectReader reader = new CsvMapper()
                .readerFor(ImportCsvDataWithBooleanPrimitive.class)
                .with(CsvSchema.emptySchema().withHeader().withArrayElementSeparator("|"))
                .with(CsvParser.Feature.EMPTY_STRING_AS_NULL)
                .with(CsvParser.Feature.FAIL_ON_MISSING_COLUMNS);

        try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(input.getBytes(StandardCharsets.UTF_8))))) {
            String header = bufferedReader.readLine();

            if (header == null || header.trim().isEmpty()) {
                errorConsumer.accept(List.of("Unable to parse empty CSV file"));
                return result;
            }

            int lineNumber = 1; //header is first line.
            String line;
            while ((line = bufferedReader.readLine()) != null) {
                lineNumber++;

                try(MappingIterator<ImportCsvDataWithBooleanPrimitive> objectMappingIterator = reader.readValues(header + "\n" + line)) {
                    List<ImportCsvDataWithBooleanPrimitive> data = objectMappingIterator.readAll().stream().toList();
                    result.addAll(data);
                } catch (CsvReadException e) {
                    String message = e.getMessage();
                    if (message.startsWith("Empty header line: can not bind data")) {
                        errorConsumer.accept(List.of("Unable to parse empty CSV file"));
                        return result;
                    }

                    if (message.startsWith("Not enough column values") || message.startsWith("Too many entries")) {
                        addErrorAtLine(parsingErrors, lineNumber, message.replaceFirst("\n.*", ""));
                    } else {
                        addErrorAtLine(parsingErrors, lineNumber, message);
                    }
                } catch (UnrecognizedPropertyException e) {
                    errorConsumer.accept(List.of("CSV contains unknown column: '%s'".formatted(e.getPropertyName())));
                } catch (InvalidFormatException e) {
                    addErrorAtLine(parsingErrors, lineNumber, e.getMessage().replaceFirst("\\n.*", ""));
                } catch (Exception e) {
                    addErrorAtLine(parsingErrors, lineNumber, e.getMessage());
                }
            }
        }  catch (Exception e) {
            errorConsumer.accept(List.of("Error reading CSV."));
        }

        if (!parsingErrors.isEmpty()) {
            errorConsumer.accept(parsingErrors);
            return result;
        }

        return result;
//        List<String> validationErrors = doBeanValidation(result);
//        if (validationErrors.isEmpty()) {
//            return result;
//        } else {
//            errorConsumer.accept(validationErrors);
//        }
    }

    private static void addErrorAtLine(List<String> parsingErrors, int lineNumber, String message) {
        parsingErrors.add(String.format("#%d: %s", lineNumber, message));
    }
```

**Reply by cowtowncoder** [reply]
Yes, you absolutely need to read lines 1-by-1 and not use "readAll()" since it does stop on first error (like it should; it's a simple convenience method). Due to variety of failure modes and so on, I am not sure there is much value in trying to provide higher level abstraction, but I am open to suggestions -- my main concern is that fundamentally only some failures are recoverable.

But what I don't get is why you seem to insist on calling "readAll()" and not -- as iterators are designed to -- use loop like

```
while (it.hasNextValue()) {
   try {
     ImportCsvDataWithBooleanPrimitive value = it.nextValue();
   } catch (...) {
   }
```

Doing that would simplify code. Similarly, header-line problems, if any, should be thrown when `MappingIterator` is created (on `ObjectReader.readValues()`), before iterator is created (I may be wrong here,  but ideally).

I hope this helps iterate to cleaner solution.

**Reply by alfonz19** [reply]
Maybe I'm overlooking something, but it didn't manage to get ti working with just an iterator. Can you show a piece of code how to create such iterator when header is read from first line? This is the reason for readAll calls I'm using. I'm artificially constructing 2 line long csv, header and line, which I read as a whole, because I didn't make it work like: initialize iterator based on header on first line, read lines using iterator.

UPDATE:
I'm reading into DTO, and I need to read header, because column order need not to be the same as field order in DTO. But if I read it like this, the line with `MappingIterator<Object> objectMappingIterator = reader.readValues(line);` fails beause header cannot be mapped to DTO.

```
@Test
    void test_withoutReadAll()  {
        String input = """
                "A","B"
                "valid1", true
                "invalid1", true, "this", "is", "too", "much!"
                "invalid2", "false"
                "invalid3"
                """;
        
        ObjectReader reader = new CsvMapper()
                .readerFor(ImportCsvDataWithBooleanPrimitive.class)
                .with(CsvSchema.emptySchema().withHeader().withArrayElementSeparator("|"))
                .with(CsvParser.Feature.EMPTY_STRING_AS_NULL)
                .with(CsvParser.Feature.FAIL_ON_MISSING_COLUMNS);

        try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(input.getBytes(StandardCharsets.UTF_8))))) {
            String line = bufferedReader.readLine();
            MappingIterator<Object> objectMappingIterator = reader.readValues(line);
            Object next = objectMappingIterator.next();
            System.out.println(next);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
```

**Reply by cowtowncoder** [reply]
I do not get your point on trying to combine header line with each individual data line.
It makes no sense to me, unless it is for trying to add more recoverability for errors (`MappingIterator` might not be able to recover from all input failures)?

Aside from that: yes, you need to catch an exception from `ObjectReader.readValues()` which constructs `MappingIterator` and fails if header line is invalid. If it does not throw exception, header handling has succeeded and iteration can continue.

On example above: that won't work because each call to `readValues()` starts from scratch, not having header line. But you should not need to do line splitting, create multiple `MappingIterator`s.

**Reply by alfonz19** [reply]
1. why I do header joining: "I have no idea how to do it differently". I know it's bad, I just don't see a path in available api.
2. Why does my last example fail — because it does start from scratch -> yes, I know that, I don't know how to do it so that it does _not_ start from scratch for each line. That's the reason for nasty line joining I resorted to. I have no idea, didn't see any mention in documentation or any sample anywhere, which can does it differently.

This is what I'm asking.

You said I can create multiple mapping iterators. Sure, but how so that it would change the situation? I can see several 'read' methods on ObjectReader, which 'will all start from scratch' as shown to fail my last example. I don't see any way in which creating multiple of them would be any game changer solving the issue.

Can you show some excerpt of code, consisting of

1. initialize, read header and eventually catch all header-related exceptions
2. loop iterating over all lines, getting new DTO or line-related exception for each iteration.

or point me to any source showing more than basic flow shown in `https://cowtowncoder.medium.com/reading-csv-with-jackson-c4e74a15ddc1`
Thanks.

**Reply by cowtowncoder** [reply]
No, I meant to say "why multiple; just create one and use that".

I do not have time to write example code at this point: iterating over values with `nextValue()`, catching exceptions is the way to go.
Good luck!

**Reply by alfonz19** [reply]
But exactly this is what I did in last example! You said that `MappingIterator.readValues` obviously fail because it reads header instead values, but there is no other way to create MappingIterator so that header is initialized from header line. Sure, calling next/nextValue and catching exception is trivial and I have NO problem doing that, just like person without apple isn't hungry because he does not know how to chew... Problem is initializing MappingIterator.

I spend 2 days on it, and I don't think it is possible or obvious how to do that. I don't ask you for whole example. I ask for at most 5 lines of code, how to initialize MappingIterator using read first header line, so that calls to next/nextValue actually works.

We spent a lot time on it both of us, I can finish full example and prepare/add it to example with comments so that others can benefit from it. But at this time it is either not possible or totally not obvious how to do it. If you cannot give more time, we will have to leave it unresolved, I cannot finish it myself correctly neither for me, nor for others as example.

**Reply by cowtowncoder** [reply]
Ok: just to make sure I understand -- and thank you for your patience here, apologies for being blunt earlier -- I still struggle understanding the problem. Given input that you have:

1. When creating `MappingIterator` for content, using `ObjectReader` with schema that has `CsvSchema.emptySchema().withHeader()`, attempt is made to read header row.
2. Reading header either:
      a. Fails -> exception, no further processing possible (it would fail for every row anyway with your work around
      b. Succeeds -> `MappingIterator` now ready for use
3. If we have iterator, every `nextValue()` should similarly either work and return value OR fail with exception (no value
4. ... but in case of Exception, iterator should in most cases recover to allow further iteration (recovery not always possible)

And I'll add code as I see it in next comment.

**Reply by cowtowncoder** [reply]
So, something like:

```
ObjectReader reader = new CsvMapper()
                .readerFor(ImportCsvDataWithBooleanPrimitive.class)
                .with(CsvSchema.emptySchema().withHeader().withArrayElementSeparator("|"))
                .with(CsvParser.Feature.EMPTY_STRING_AS_NULL)
                .with(CsvParser.Feature.FAIL_ON_MISSING_COLUMNS);
MappingIterator<ImportCsvDataWithBooleanPrimitive> it;
try {
   it = reader.readValues();
} catch (IOException e) {
    throw new RuntimeException("Bad header row, root cause: "+e.getMessage(), e);
} 

while (it.hasNextValue()) {
  try {
     ImportCsvDataWithBooleanPrimitive value = it.nextValue();
     // process value
  } catch (IOException e) {
     // or handle problem
     ++errorCount;
  } 
}
```

Hope this helps.

**Reply by alfonz19** [reply]
TLDR: Yes, it helped me a lot, fixed. Thank you very much.

Uff, now I get it. Apologies, I have no idea how could I overlook the possibility of passing whole csv data into com.fasterxml.jackson.databind.ObjectReader#readValues(java.lang.String). I overcomplicated it for myself with BufferedReader and then was surprised that it does not work. Thank you **very much** for your help.

full adapted test with assertions with same input as we had it before looks like following:

```
@Test
    void testReadingWithCollectingErrorPerLine() {
        String input = """
                "A","B"
                "valid1", true
                "invalid1", true, "this", "is", "too", "much!"
                "invalid2", "false"
                "invalid3"
                """;
        List<String> errors = new ArrayList<>();
        List<ImportCsvDataWithBooleanPrimitive> values = new ArrayList<>();

        ObjectReader reader = new CsvMapper()
                .readerFor(ImportCsvDataWithBooleanPrimitive.class)
                .with(CsvSchema.emptySchema().withHeader().withArrayElementSeparator("|"))
                .with(CsvParser.Feature.EMPTY_STRING_AS_NULL)
                .with(CsvParser.Feature.FAIL_ON_MISSING_COLUMNS);
        try(MappingIterator<ImportCsvDataWithBooleanPrimitive> it = reader.readValues(input)) {
            int lineNumber = 1;
            while (it.hasNextValue()) {
                lineNumber++;
                try {
                    values.add(it.nextValue());
                } catch (JacksonException e) {
                    errors.add(String.format("#%d:%d %s", lineNumber, e.getLocation().getColumnNr(), e.getOriginalMessage()));
                } catch (Exception e) {
                    errors.add(String.format("#%d %s", lineNumber, e.getMessage()));
                }
            }
        } catch (IOException e) {
            throw new RuntimeException("Bad header row, root cause: "+e.getMessage(), e);
        }


        assertThat(values.size(), is(1));
        assertThat(errors.size(), is(3));

        assertThat(values.getFirst(), is(new ImportCsvDataWithBooleanPrimitive("valid1", true)));

        String expected = """
                #3:18 Too many entries: expected at most 2 (value #2 (6 chars) ""this"")
                #4:12 Cannot deserialize value of type `boolean` from String ""false"": only "true"/"True"/"TRUE" or "false"/"False"/"FALSE" recognized
                #5:1 Not enough column values: expected 2, found 1""";
        assertThat(String.join("\n", errors), is(expected));
    }
```

if you would like to add it to your how-to page, with test cases covering bad header etc. I will expand it and finish it to cover all cases. Would you like me to do that 'for you' and others? I can put there better input demonstrating also array splitting. Just tell me; you helped me, I can create demonstration of this which would look like you want it. And please advise where to put it if you want me to create MR, otherwise I can put it here.

**Reply by cowtowncoder** [reply]
@alfonz19 Great! Glad it was just bit of misunderstanding, b/w our assumptions.

On improvement:  I appreciate the offer & think this would be great addition to `csv/README.md` (as seen at https://github.com/FasterXML/jackson-dataformats-text/tree/2.19/csv ), along with any fixes? Change would be done by a simple pull request.

**Reply by alfonz19** [reply]
(please be patient with me; I have some deadline quite close which would cost me some money. I'd either have to rush it, which I don't want or I have to try to do it a little later. I will definitely try to do that, probably in a week. Sorry for delay).

**Reply by cowtowncoder** [reply]
@alfonz19 No problem at all. Take your time.
